### ISSUE TITLE
Golang 1.9.2 and above fails in epoll

### ISSUE THREAD

**Issue by wkozaczuk** [cloudius-systems/osv#971]
Even after implementing SYSCALL switch stack functionality golang-httpserver app fails consistently under enough load with following error:

ab -c 10 -n 100 http://localhost:8000/

```
-->netpoll: START
-->netpoll: START
Go version: go1.9.2
-->netpollinit: START
-->netpollopen: START
-->netpoll: START
-->netpoll: START
-->netpollopen: START
-->netpollopen: START
-->netpollopen: START
-->netpollopen: START
-->netpoll: START
-->netpoll: START
-->netpoll: START
-->netpollopen: START
-->netpoll: START
-->netpollopen: START
-->netpollopen: START
-->netpollopen: START
-->netpoll: START
-->netpoll: START
-->netpoll: START
-->netpoll: START
-->netpoll: START
-->netpollopen: START
-->netpoll: START
-->netpoll: START
-->netpollopen: START
-->netpoll: START
-->netpollopen: START
-->netpollopen: START
-->netpollopen: START
-->netpollopen: START
-->netpollopen: START
-->netpollopen: START
-->netpollopen: START
-->netpollopen: START
-->netpollopen: START
-->netpollopen: START
-->netpollopen: START
-->netpollopen: START
-->netpollopen: START
-->netpollopen: START
-->netpollopen: START
-->netpollopen: START
Assertion failed: !(e & ~SUPPORTED_EVENTS) (core/epoll.cc: events_poll_to_epoll: 53)

[backtrace]
0x0000000000225a88 <__assert_fail+24>
0x0000000000405231 <epoll_file::wait(epoll_event*, int, int)+2641>
0x0000000000402d30 <epoll_wait+112>
0x00000000003d7725 <syscall+1013>
0x00000000003d8023 <syscall_wrapper+83>
0x0000000000389d9d <syscall_entry+157>
0x0000100000d6b468 <???+14070888>
0x0000100000d42cf0 <???+13905136>
0x0000100000d4368f <???+13907599>
0x0000100000d440fb <???+13910267>
0x0000100000d67172 <???+14053746>
0x0000100000d69da5 <???+14065061>
```

**Comment by wkozaczuk** [comment]
Here is what I got:

```
-->netpollopen: START
-->netpollopen: START
-->netpollopen: START
-->netpollopen: START
unexpected return from poll 0x20 - 0x20

[backtrace]
0x0000000000405226 <epoll_file::wait(epoll_event*, int, int)+2630>
0x0000000000402d30 <epoll_wait+112>
0x00000000003d7725 <syscall+1013>
0x00000000003d8023 <syscall_wrapper+83>
0x0000000000389da1 <syscall_entry+161>
0x0000100000d6b468 <???+14070888>
0x0000100000d42cf0 <???+13905136>
0x0000100000d4368f <???+13907599>
0x0000100000d439ab <???+13908395>
0x0000100000d67172 <???+14053746>
0x0000100000d3dc5f <???+13884511>
0x0000100000d2768b <???+13792907>
0x0000100000d69da0 <???+14065056>
```

**Comment by nyh** [comment]
0x20 is POLLNVAL.

poll() returns POLLNVAL in several strange situations - one of them is when the requested bits were improper. I don't think this is the case (we check the incoming bits). Another is when the file descriptor has been closed. You can add printouts in poll.c where it returns POLLNVAL to find out if this is what happened.

According to documentation I read, epoll should never return POLLNVAL. Perhaps (I'm not 100% sure) in epoll() if we see an POLLNVAL we should drop this event, i.e., where we have in epoll.cc 

```
active = key._file->poll(events_epoll_to_poll(evt.events));
```

we can try following it with something like

```
if (active & POLLNVAL) {
    active = 0;
}
```

If you can reproduce this problem you can check if this works, and if this makes sense. In particular, do we get this problem just once, or in an endless loop (in which case maybe we also need to remove the polling on this file descriptor).

**Comment by wkozaczuk** [comment]
It actually happens in **poll_no_poll**. What is interesting is that sometimes couple of first calls the POLLNVAL gets ignored but eventually one of the results in an abort/assert in **events_poll_to_epoll**.

So the question seems to be why sometimes call to poll_no_poll() by vfs_file::poll() result in the assert in events_poll_to_epoll and sometimes not?

Here are the code changes are made to print debug:

```
diff --git a/core/epoll.cc b/core/epoll.cc
index f425c4a..6bf578d 100644
--- a/core/epoll.cc
+++ b/core/epoll.cc
@@ -50,7 +50,9 @@ inline uint32_t events_epoll_to_poll(uint32_t e)
 }
 inline uint32_t events_poll_to_epoll(uint32_t e)
 {
-    assert (!(e & ~SUPPORTED_EVENTS));
+    if (e & ~SUPPORTED_EVENTS) {
+        abort("unexpected return from poll 0x%x - 0x%x\n", e, e & ~SUPPORTED_EVENTS);
+    }
     return e;
 }

diff --git a/core/poll.cc b/core/poll.cc
index a036bac..62b2b14 100644
--- a/core/poll.cc
+++ b/core/poll.cc
@@ -74,8 +74,10 @@ int poll_no_poll(int events)
      * functionality is present without hard-coding knowledge
      * of specific filesystem implementations.
      */
-    if (events & ~POLLSTANDARD)
+    if (events & ~POLLSTANDARD) {
+        debugf("POLLNVAL: poll_no_poll: Not standard - 0x%x - 0x%x\n", events, events & ~POLLSTANDARD);
         return (POLLNVAL);
+    }

     return (events & (POLLIN | POLLOUT | POLLRDNORM | POLLWRNORM));
 }
@@ -119,6 +121,7 @@ int poll_scan(std::vector<poll_file>& _pfd)
         if (!fp) {
             entry->revents |= POLLNVAL;
             nr_events++;
+            debugf("POLLNVAL: poll_scan: No file\n");
             continue;
         }

@@ -336,6 +339,7 @@ static int poll_one(struct pollfd& pfd, file::timeout_t timeout)
     auto fref = fileref_from_fd(pfd.fd);

     if (!fref) {
+        debugf("POLLNVAL: poll_one: No file\n");
        pfd.revents = POLLNVAL;
         return 1;
     }
```

Here is the console output:
```
syscall(): unimplemented system call 267
Go version: go1.9.2
POLLNVAL: poll_no_poll: Not standard - 0x80002005 - 0x80002000

POLLNVAL: poll_no_poll: Not standard - 0x80002005 - 0x80002000
POLLNVAL: poll_no_poll: Not standard - 0x80002005 - 0x80002000
POLLNVAL: poll_no_poll: Not standard - 0x80002005 - 0x80002000
POLLNVAL: poll_no_poll: Not standard - 0x80002005 - 0x80002000
POLLNVAL: poll_no_poll: Not standard - 0x80002005 - 0x80002000
POLLNVAL: poll_no_poll: Not standard - 0x80002005 - 0x80002000
POLLNVAL: poll_no_poll: Not standard - 0x80002005 - 0x80002000
POLLNVAL: poll_no_poll: Not standard - 0x80002005 - 0x80002000
POLLNVAL: poll_no_poll: Not standard - 0x80002005 - 0x80002000
POLLNVAL: poll_no_poll: Not standard - 0x80002005 - 0x80002000
unexpected return from poll 0x20 - 0x20

[backtrace]
0x0000000000405256 <epoll_file::wait(epoll_event*, int, int)+2630>
0x0000000000402d60 <epoll_wait+112>
0x00000000003d7725 <syscall+1013>
0x00000000003d8023 <syscall_wrapper+83>
0x0000000000389da1 <syscall_entry+161>
0x0000100001d6b3f8 <???+30847992>
0x0000100001d47c42 <???+30702658>
0x0000100001d409b1 <???+30673329>
0x0000100001d40885 <???+30673029>
0x0000100001f22032 <???+32645170>
0x000020000055ff3f <???+5635903>
```

**Comment by nyh** [comment]
The "nonstandard" bits 0x80002000 given to poll_no_poll() is EPOLLET | EPOLLRDHUP.
I think it is wrong that poll_no_poll() returns POLLNVAL in this case. Looking at the poll() manual page, I don't see any indication that poll() should return POLLNVAL if unrecognized event bits are on... So it's wrong for poll_no_poll() to ever return POLLNVAL, and I don't understand the comment suggesting why it does.
I am wondering, can't poll_no_poll() just always return 0? What would be wrong with that?

Another small issue needs fixing is that the epoll.cc code should have removed the EPOLLET bit before sending it to poll().  The function events_epoll_to_poll() should  `return e & ~(EPOLLET | EPOLLONESHOT)` and not "e", so those two epoll-specific flags won't be passed to poll which doesn't know what to do with them (most poll implementations just ignored them, but apparently poll_no_poll() didn't).

**Comment by wkozaczuk** [comment]
Couple of questions:
1. Regarding your suggestion to drop POLLNVAL in epoll.cc:process_activity() - I wonder if it is valid. What if POLLNVAL was a symptom of some legitimate error? Should it somehow be reported to caller.

2. I added debug statements to poll() method of all classes extending from file and indeed 0x20 seems to be coming from vfs_file::poll() which calls poll_no_poll(). But I still do not understand why in all occurrences but final one POLLNVAL is returned by poll_no_poll() causes not harm (no crash/no assertion broken) - see my last log. Do you have good explanation for this? Based on code analysis I see poll_no_poll is only called by vfs_file::poll(). Or is golang app calling standard poll() which ends up calling  vfs_file::poll() as well but with no harm.

**Comment by nyh** [comment]
1. In my last comment, I didn't recommend to filter out POLLNVAL, just the poll_no_poll() shouldn't produce this bit.
2. I have no idea. Perhaps the epoll.cc has another code path which also polls but doesn't have this assert (e.g., for setting up the poll).

Looking again at the code, and seeing the silly-named poll_no_poll() is only used for disk files, which are always ready for read, write, etc., I think it will be safe to just drop the
```c++
    if (events & ~POLLSTANDARD)
        return (POLLNVAL);
```
from poll_no_poll(). So the file will be considered "readable" even if an irrelevant bit like POLLRDHUP is also enabled. I don't know, by the way, why the code bothers to poll a ZFS file, but if it does so, with EPOLLET, it will anyway just get one event and that's it.

You can also test on Linux what happens when you poll() a normal disk file with a non-standard bit enabled. I suspect the bit is just ignored, and POLLNVAL is not returned.

**Comment by wkozaczuk** [comment]
I think I finally understand the root cause. I might be restating what you wrote above. 

The reason why POLLNVAL returned by poll_no_poll() does not cause any harm in all but last occurrence is simply because it client (Golang program) directly calls poll() against a file. In last occurrence when it crashes the file is actually watched through epoll (which calls poll() in its implementation). So indeed what you suggested the poll_no_poll() implementation is problematic. Does you suggestion still hold given that poll() maybe called directly or through epoll()?

Some logs (with thread IDs in front):
```
-->netpollopen: START
61:-->vfs_file::poll() - 0x80002005
-->netpollopen: START
61:-->poll_no_poll: POLLNVAL: - 0x80002005 - 0x80002000

56:-->vfs_file::poll() - 0x80002005
56:-->poll_no_poll: POLLNVAL: - 0x80002005 - 0x80002000
-->netpoll: START
56:epoll_file::process_activity(): BEFORE file->poll(): [0x80002005 -> 0x80002005]
-->netpoll: START
56:socket_file::poll: events - [0x80002005], sopoll - [0x1]

56:epoll_file::process_activity(): active from file->poll(): [0x80002005 -> 0x1]
-->netpollopen: START
60:-->vfs_file::poll() - 0x80002005
60:-->poll_no_poll: POLLNVAL: - 0x80002005 - 0x80002000
....
/// AND the last one
....
61:epoll_file::process_activity(): BEFORE file->poll(): [0x80002005 -> 0x80002005]
55:socket_file::poll: events - [0x80002005], sopoll - [0x5]
61:-->vfs_file::poll() - 0x80002005
-->netpollopen: START
61:-->poll_no_poll: POLLNVAL: - 0x80002005 - 0x80002000

61:epoll_file::process_activity(): active from file->poll(): [0x80002005 -> 0x20]
-->netpollopen: START
unexpected return from poll 0x20 - 0x20

```

The reason this is a bug and was never discovered is because our tst-epoll tests only against pipe not files nor sockets.

**Comment by nyh** [comment]
I have patches for the underlying problems, and I'll send them in a few minutes, but I still don't understand how to test it:
In Linux, you can't epoll an ordinary file -  EPOLL_CTL_ADD fails. (in OSv, this steps succeeds, which we should probably fix, but it's not important now).
So we can only test this on sockets, and I see in your last comment that this was indeed a socket, not a file.
However, sockets don't normally get to poll_no_poll - they call sopoll(), and I wonder in which situation does sopoll call vfs_file::poll(), and why. That may also be a bug.

**Comment by wkozaczuk** [comment]
As you are suggesting poll() or epoll* against regular files is weird but my log outputs show exactly that (besides socket files). 

```
60:-->vfs_file::poll() - 0x80002005
60:-->poll_no_poll: POLLNVAL: - 0x80002005 - 0x80002000
```

The vfs_file::poll() logs this. 

BTW I enabled log to see where sys_open() is called (which constructs vfs_file()) and here is a file from procfs:

```
sysopen(): /proc/sys/kernel/hostname]
```

**Comment by wkozaczuk** [comment]
I just found the note that "files" (nor really files) under /proc are pollable on Linux. So maybe procfs is not implemented correctly in OSv?

**Comment by nyh** [comment]
That's also what I thought, but I wrote a test, and Linux does *not* allow epoll against a regular file on ext4 filesystem (the EPOLL_CTL_ADD operation fails), so because this code also works on Linux, I don't see how it can be a disk file. But maybe a file in /proc is different! I'll test again.

**Comment by nyh** [comment]
Indeed, you are right - files under /proc are indeed pollable on Linux. So just sent to the mailing list a patch with a test for polling a /proc file (in OSv, this uses poll_no_poll()) and checking that extra bits like POLLRDHUP do not cause a crash (after the patches I already sent).
@wkozaczuk please test / review my patches. Thanks!

**Comment by wkozaczuk** [comment]
I think your patches are correct and the test is passing. However the buggy behavior in poll_no_poll() which you fixed unintentionally made polling on /proc/sys/kernel/hostname work. However now Golang httpserver app only responds to first http request. Second one (which probably triggers poll/epoll) does not respond and I am guessing is caused by Golang to wait for an event on /proc/sys/kernel/hostname which never comes. This is just a theory - I did not apply your patches to my branch that has debug statements,

**Comment by nyh** [comment]
On Thu, May 17, 2018 at 4:01 PM, WALDEMAR KOZACZUK <notifications@github.com
> wrote:

> I think your patches are correct and the test is passing. However the
> buggy behavior in poll_no_poll() which you fixed unintentionally made
> polling on /proc/sys/kernel/hostname work.
>
It wasn't unintentially - this was the whole intention of this change!
So I'll commit these patches.

However now Golang httpserver app only responds to first http request.
> Second one (which probably triggers poll/epoll) does not respond and I am
> guessing is caused by Golang to wait for an event on
> /proc/sys/kernel/hostname which never comes. This is just a theory - I did
> not apply your patches to my branch that has debug statements,
>
I think you would need to debug further, I am doubtful about your theory
because why would the *second* request want to read from
/proc/sys/kernel/hostname? I don't know why they ever polled on this file,
but after already having read it for the first request, why read it again?
Do they check the hostname on every request? That would be very strange.

I saw in your previous printouts that epoll was used with EPOLLET (edge
triggered). Maybe we have a bug on how this mode works, and we don't get
the second event on the listening socket (EPOLLET is not the same as
EPOLLONESHOT, we should get a second event after having processed the first
one).

**Comment by wkozaczuk** [comment]
Sorry for not opening new issue because I still think it is related to this
one or at least the patches you have sent.

I only noticed now that you made this change in events_epoll_to_poll:

-    return e;
+    return e & SUPPORTED_EVENTS;

when I reverted to the original version (just returning e) my "hangs"
problem goes away and golang httpserver finally works fine on 1.9.2.

At first I thought it might be because SUPPORTED_EVENTS is of type int
whereas e and return type of events_epoll_to_poll is uint32_t. But
changing SUPPORTED_EVENTS
to uint32_t did not help.

I still do not understand why reverting to 'return e;' helps because anding
with SUPPORTED_EVENTS makes sense but somehow causes golang httpserver to
hang after first http request.

**Comment by nyh** [comment]
Sorry, for botching this up. I'll send a new patch for #974 that will explain what went wrong. As you already figured out there, the problem is EPOLLET.

**Comment by nyh** [comment]
Thanks. You can test my (indeed trivial) patch, but I didn't add any unit test for this (namely, checking sockets and EPOLLET for the missed wakeup bug, that was originally fixed in d41d748f89cf6e47820e8585cef6eedb837c05be but needs to know the EPOLLET bit which I botched up in commit 9eb6f2e38dc0103bd0d64a17221e2cc59b8cb138). If you have the energy to write a test for this, it would be great :-)
